src/raven/client.clj

```diff
@@ -209,30 +214,35 @@
        ;; Make sure we clear the breadcrumbs from the thread-local storage.
        (clear-breadcrumbs))))
   ([dsn ev]
-   (capture! (http/build-client {}) dsn ev)))
-
-(defn make-breadcrumb
-  "Create a breadcrumb map."
-  [message category level timestamp]
-  ;; TODO: Extend to support more than just default breadcrumbs.
-  {:type "default"
-   :timestamp timestamp
-   :level level
-   :message message
-   :category category}
+   (capture! @@thread-storage dsn ev)))
+
+(defn make-breadcrumb!
+  "Create a breadcrumb map.
+    
+  level can be one of:
+    ['debug' 'info' 'warning' 'warn' 'error' 'exception' 'critical' 'fatal']"
+  ([message category]
+   (make-breadcrumb! message category "info"))
+  ([message category level]
+   (make-breadcrumb! message category level (timestamp!)))
+  ([message category level timestamp]
+   {:type "default" ;; TODO: Extend to support more than just default breadcrumbs.
+    :timestamp timestamp
+    :level level
+    :message message
+    :category category})
   ;; :data (expected in case of non-default)
 )
 
 (defn add-breadcrumb!
-  "Append a breadcrumb to the list of breadcrumbs for this thread.
+  "Append a breadcrumb to the list of breadcrumbs.
 
-  level can be one of:
-    ['debug' 'info' 'warning' 'warn' 'error' 'exception' 'critical' 'fatal']
+  The context is expected to be a map, and if no context is specified, a
+  thread-local storage map will be used instead.
   "
-  ([message category]
-   (add-breadcrumb! message category "info"))
-  ([message category level]
-   (add-breadcrumb! message category level (timestamp!)))
-  ([message category level timestamp]
-   ;; We need to dereference to get the atom since "breadcrumbs" is thread-local.
-   (swap! @breadcrumbs #(conj % (make-breadcrumb message category level timestamp)))))
+  ([breadcrumb]
+   ;; We need to dereference to get the atom since "thread-storage" is thread-local.
+   (swap! @thread-storage (fn [x] (update x :breadcrumbs conj breadcrumb))))
```

Comment:
I don't understand how this works.

"(swap! @thread-storage" -> swap the value of the thread-storage atom
"add-breadcrumb! breadcrumb)" -> swap in the value of a recursive call to add-breadcrumb! with the same parameter (?)

This cannot produce the atom I need {:breadcrumbs [{bc1..} {bc2}]}. This should blow the stack because it's non-tail recursion (or at least it looks like it).

However my tests still pass. I don't understand.

Reply:
Does swap! add the  current value of the atom *between* the function and the breadcrumb parameter? As in, doing ((partial add-breadcrumb! @@thread-storage) breadcrumb)?

Reply:
Answering my own question: yes, that is exactly what is happening here.

It's very counter-intuitive (at least for the clojure beginner that I am), but it works.

Reply:
Yes this is the case.
Clojure uses the idiom of *sending* a function to a value in a number of constructs.
Notably, this is the case with: `swap!`, `vswap!`, `update`, `send!`, and `alter-var-root` on the top of my head. This should read send function `f` on value `v taking optional arguments`:

```clojure
(def v {:a 0})

(update v :a inc) ;; => {:a 1}
(update v :b conj 0) ;; => {:a 0 :b (0)}
```